api/include/mfxstructures.h

```diff
@@ -328,7 +328,8 @@ typedef struct {
             mfxU16  SliceGroupsPresent;
             mfxU16  MaxDecFrameBuffering;
             mfxU16  EnableReallocRequest;
-            mfxU16  reserved2[7];
+            mfxU16  EnableZeroNumFrameActual;
```

Comment:
I'm afraid API change is not so simple, you need to
1) get approval from @varistar 
2) add static assert to [this](https://github.com/Intel-Media-SDK/MediaSDK/blob/master/_studio/shared/src/mfx_static_assert_structs.cpp) file
3) add description to doc with example on when we need to use this control

Reply:
@mgonchar thanks for the review, fixed. 
I will do the changes after aligning the direction with @varistar . 

Hi @varistar 
What's your suggestion on this?

thanks

Reply:
Hi, several questions:
1. Why we cannot remove AllocFrames from Decoders' initialization at all instead of making fake AllocFrames with zero output? The complexity of changes or minimize Linux specific or what? The current solution looks like a workaround to me. 
2. Who will care about close source Win changes? This mode should be prohibited for DX9, but can be enabled in (or not) DX11?

Reply:
Hi @varistar 
thanks for the reply.

>Why we cannot remove AllocFrames from Decoders' initialization at all instead of making fake AllocFrames with zero output? The complexity of changes or minimize Linux specific or what? The current solution looks like a workaround to me.

Yes, I'd love to remove AllocFrames. But it may introduce back capability issues for old MediaSDK users. So I just provide this workaround.

> Who will care about close source Win changes? This mode should be prohibited for DX9, but can be enabled in (or not) DX11?

It's a design flaw of DX9, DX11 removed this constrain. 
Win7 is the last windows uses DX9. I hope we can use this mode for both windows and Linux after Win7 retired.

Reply:
You are right, we cannot just remove AllocFrames, but we can do the following - 
if (!mfxVideoParam->lazy_allocation)
AllocFrames()
....
else
// no call of AllocFrames.

Right now anybody who wants to support this mode needs to create very special allocator which allocates zero surfaces that looks very cont-intuitive. 
Ideally, I'd like to see a solution where applications don't need to set external allocator and MediaSDK doesn't call AllocFrames/Free when it's not needed

Reply:
> You are right, we cannot just remove AllocFrames, but we can do the following -
> if (!mfxVideoParam->lazy_allocation)
> AllocFrames()
> ....
> else
> // no call of AllocFrames.
> 

Yes. we are fully aligned. 
This is exactly I suggested in https://github.com/Intel-Media-SDK/MediaSDK/pull/2145#issuecomment-644501955
I will send a patch for this.
Before the patch, could you share your preferred name? SkipFrameAllocFree or LazyAllocation in mfxInfoMFX?

> Ideally, I'd like to see a solution where applications don't need to set external allocator and MediaSDK doesn't call AllocFrames/Free when it's not needed

This may need more works in MediaSDK. Because MeidaSDK needs mfxFrameAllocator.getHDL to get platform related surface handle.

Reply:
github is broken last week. @varistar made some valuable suggestions through mail. let me attach here

replays from @varistar 

>  > This may need more works in MediaSDK. Because MeidaSDK needs mfxFrameAllocator.getHDL to get platform related surface handle.
>  
>  Agree, but for such mode we can specify that mfxMemID should contain native handle making getHDL call redundant.
>  
> 
>  
> >  Before the patch, could you share your preferred name? SkipFrameAllocFree or LazyAllocation in mfxInfoMFX?
>  
>  I'd like to call it like DeferredFrameAllocation or DelayedFrameAllocation
>

Reply:
>Agree, but for such mode we can specify that mfxMemID should contain native handle making getHDL call redundant.

Done. But allocator is a session level param. It can't controlled by mfxVideoParam. So I use     `MFXVideoCORE_SetFrameAllocator(m_session, NULL);`
to indicate a delayed frame allocator.
refer to https://github.com/Intel-Media-SDK/MediaSDK/pull/2145/files#diff-fae1222e835b7a0eea6af66261c4d4d0R904 fro implementation details

The patch tested for following test cases:
**decode**: vp9 resolution change at P frame.
**transcode with vpp**: h264->h264, h264->h265

@mgonchar @dmitry-gurulev @varistar , please help  review again.

thanks